compiler/modules/front.syntax/syntaxNode.bal

```diff
@@ -620,3 +621,273 @@ function flattenSyntaxNodeList((SyntaxNode[]|SyntaxNode?)[] arr) returns SyntaxN
     }
     return nodes;
 }
+
+public function exprToString(Expr expr) returns string {
+    return "\n".'join(...syntaxNodeToString(syntaxNodeFromExpr(expr)));
+}
+
+// join words without space
+const CLING = ();
+// line feed
+const LF = 0;
+// line feed and indent
+const LF_INDENT = 1;
+// line feed and outdent
+const LF_OUTDENT = -1;
+
+type Word string|LF_INDENT|LF_OUTDENT|LF|CLING;
+
+function syntaxNodeToString(SyntaxNode|RootSyntaxNode node) returns string[] {
+    Word[] words = [];
+    syntaxNodeToWords(words, node);
+    return concat(...words);
+}
+
+function syntaxNodeToWords(Word[] words, SyntaxNode|RootSyntaxNode node) {
+    if node is TerminalSyntaxNode {
+        words.push(terminalSyntaxNodeToString(node));
+    }
+    else {
+        AstNode? astNode = node !is RootSyntaxNode ? node.astNode : ();
```

Comment:
This is making stringNodeToString more tightly coupled to the AST than it should be.

I suggest you come up with a set of flags that describe the various differences in how the child nodes of non-terminal need to be formatted.  Set the flags based on the AST node. Then use the flags to output the list of child nodes.

The indent/outdent doesn't feel good to me.  Why can't you just do it based on `{` and `}` terminals?

Reply:
fixed